### ISSUE TITLE
routecomponent's authentication functions do not work as expected

### ISSUE THREAD

**Issue by spearec** [OpenEnergyDashboard/OED#817]
### Describe the bug

Due to how asynchronous functions are handled, the current method in which checkAuth, requireAuth, and requireRole are written has a fundamental issue. The token validation is an asynchronous function, but the functions themselves must return a value synchronously. This means a value is returned before the async block is finished. because of this, users are directed to the page they try to access regardless of the token authentication. This is the root cause of issue #816, although I have found a workaround for that issue.

Currently, apart from #816, this is a silent issue. Some error is encountered during the process which causes the token to be removed and then the function to be re-run, which will properly redirect the user (see screenshots). For now, this works as intended. Users are able to access a page only if they have access, and invalid tokens are deleted and then the user is redirected. The problem is that the code only works because of further errors popping up downstream.

For example, consider requireAuth:
```typescript
public requireAuth(component: JSX.Element) {
// Redirect route to home page if the auth token does not exist
if (!hasToken() || !this.props.loggedInAsAdmin) {
	return <Redirect to='/' />;
}
// Verify that the auth token is valid.
// Needs to be async because of the network request
(async () => {
	if (!(await verificationApi.checkTokenValid())) {
		showErrorNotification(translate('invalid.token.login.admin'));
		// We should delete the token when we know that it is expired. Ensures that we don't not leave any unwanted tokens around.
		deleteToken();
		// This ensures that if there is no token then there is no stale profile in the redux store.
		this.props.clearCurrentUser();
		// Route to login page since the auth token is not valid
		return <Redirect to='/' />;
	}
	return component;
})();
return component;
}
```

When the function is evaluated, if the user has a token, the last return statement is immediately returned. Only after the await function finishes does the rest of the code evaluate. This means that the function may attempt to return a redirect, but this second return value is never processed.

### To Reproduce

1. Log in as an admin user
2. Wait for the token to become invalid
3. Access an admin page
4. Inspect the web console

### Expected behavior

There should be no improper API calls nor other react errors.

### Screenshots

![image](https://user-images.githubusercontent.com/113216142/202008052-43c1afcb-89b8-4f32-8fe9-928100452882.png)

**Comment by ChrisMart21** [comment]
This issue is currently being looked into and will be resolved in an upcoming migration to Redux-Toolkit, and react-router-v6

**Comment by huss** [comment]
Fixed by PR #1113 so closing.
